### PR TITLE
Segregated witness support

### PR DESCRIPTION
**Pull request by luke-jr**

### CODE REVIEW

**Comment by gmaxwell** [comment]
This should compare to the GBT provided commitment if there is one and if the txn list is the same, as a sanity check. no?

File: bitcoin/txn.py [hunk 1]
```diff
@@ -52,6 +55,9 @@ def addOutput(self, amount, pkScript):
 		self.outputs.append( (amount, pkScript) )
 	
 	def disassemble(self, retExtra = False):
+		if self.data[4:6] == b'\0\1':
```
Comment (theuni) [inline bitcoin/txn.py:58]: Won't this cause issues for pre-0.13 nodes mining anyonecanspends?

File: merklemaker.py [hunk 2]
```diff
@@ -82,6 +94,7 @@ def __init__(self, *a, **k):
 		self.currentBlock = (None, None, None)
 		self.lastBlock = (None, None, None)
 		self.SubsidyAlgo = lambda height: 5000000000 >> (height // 210000)
+		self.WitnessNonce = b'\0' * 0x20
```
Comment (theuni) [inline merklemaker.py:117]: stick this in a constant somewhere?

File: bitcoin/txn.py (hunk 1 above)
Comment (luke-jr) [reply bitcoin/txn.py:58]: Why would it? A transaction with the witness dummy/flags should be entirely invalid pre-segwit.

File: merklemaker.py (hunk 2 above)
Comment (luke-jr) [reply merklemaker.py:117]: This is essentially a constant.

**Comment by luke-jr** [comment]
@gmaxwell Maybe. Block proposals seem to be a superset of that comparison though?

File: bitcoin/txn.py [hunk 3]
```diff
@@ -115,7 +124,21 @@ def assemble(self):
 		self.idhash()
 	
 	def idhash(self):
+		if self.data[4:6] == b'\0\1':
```
Comment (pstratem) [inline bitcoin/txn.py:127]: This needs a comment

File: eloipool.py [hunk 4]
```diff
@@ -131,6 +134,10 @@ def makeCoinbaseTxn(coinbaseValue, useCoinbaser = True, prevBlockHex = None):
 	pkScript = BitcoinScript.toAddress(config.TrackerAddr)
 	txn.addOutput(coinbaseValue, pkScript)
 	
+	# SegWit commitment
+	if not witness_commitment is None:
```
Comment (pstratem) [inline eloipool.py:138]: if witness_commitment is not None:

File: bitcoin/script.py [hunk 5]
```diff
@@ -43,6 +45,13 @@ def toAddress(cls, addr):
 		elif ver == 5 or ver == 196:
 			return b'\xa9\x14' + pubkeyhash + b'\x87'
 		raise ValueError('invalid address version')
+
```
Comment (pstratem) [inline bitcoin/script.py:48]: Does this really belong in BitcoinScript?

File: bitcoin/txn.py (hunk 1 above)
Comment (pstratem) [reply bitcoin/txn.py:58]: This is fine but like the other use of this below needs a comment.

File: bitcoin/script.py (hunk 5 above)
Comment (luke-jr) [reply bitcoin/script.py:48]: Yes, BitcoinScript simply creates standard transaction formats from params.

**Comment by btcdrak** [comment]
Tested ACK 3aab63ecee5e35f259e4b1aa0bfe819de39c95bd 

https://www.blocktrail.com/tBTC/block/0000000000006d4bf2ec7a114c12285a6d1049c2634d81cfb903c9c8c9a3a9af
https://www.blocktrail.com/tBTC/block/00000000000004028f4f8d73bf04adab032dd634d460ba642f88e1fa84ff077e

**Pull request merged**